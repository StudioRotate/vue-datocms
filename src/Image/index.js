import PropTypes from "@znck/prop-types";

const absolutePositioning = {
  position: "absolute",
  left: 0,
  top: 0,
  width: "100%"
};

const Image = {
  props: {
    data: PropTypes.shape({
      aspectRatio: PropTypes.number.isRequired,
      width: PropTypes.number.isRequired,
      base64: PropTypes.string,
      height: PropTypes.number,
      sizes: PropTypes.string,
      src: PropTypes.string,
      srcSet: PropTypes.string,
      webpSrcSet: PropTypes.string,
      bgColor: PropTypes.string,
      alt: PropTypes.string,
      title: PropTypes.string
    }).isRequired,
    pictureClassName: PropTypes.string,
    fadeInDuration: PropTypes.number,
    intersectionTreshold: PropTypes.number.defaultValue(0),
    intersectionMargin: PropTypes.string.defaultValue("0px 0px 0px 0px"),
    lazyLoad: PropTypes.bool.defaultValue(true),
    pictureStyle: PropTypes.object
  },
  inheritAttrs: false,
  data: () => ({
    observer: null,
    inView: false,
    loaded: false
  }),
  computed: {
    addImage() {
      if (!this.lazyLoad) {
        return true;
      }

      if (typeof window === "undefined") {
        return false;
      }

      if ("IntersectionObserver" in window) {
        return this.inView || this.loaded;
      }

      return true;
    },
    showImage() {
      if (!this.lazyLoad) {
        return true;
      }

      if (typeof window === "undefined") {
        return false;
      }

      if ("IntersectionObserver" in window) {
        return this.loaded;
      }

      return true;
    }
  },
  methods: {
    load() {
      if (this.$el.getAttribute("src") !== this.srcPlaceholder) {
        this.loaded = true;
      }
    }
  },
  render(h) {
    const {
      data,
      fadeInDuration,
      pictureClassName,
      pictureStyle,
      showImage,
      addImage
    } = this;

    const webpSource =
      data.webpSrcSet &&
      h("source", {
        attrs: {
          srcset: data.webpSrcSet,
          sizes: data.sizes,
          type: "image/webp"
        }
      });

    const regularSource =
      data.srcSet &&
      h("source", { attrs: { srcset: data.srcSet, sizes: data.sizes } });

    const placeholder = h("div", {
      style: {
        paddingTop:
          data.width && data.height
            ? `${(data.height / data.width) * 100.0}%`
            : `${100.0 / data.aspectRatio}%`,
        backgroundImage: data.base64 ? `url(${data.base64})` : null,
        backgroundColor: data.bgColor,
        backgroundSize: "cover"
      }
    });

    console.log('INVIEW', this.inView);

    return h(
      "div",
      {
        style: {
          display: "inline-block",
          maxWidth: "100%",
          width: `${data.width}px`,
          overflow: "hidden",
          position: "relative"
        }
      },
      [
        placeholder,
        addImage &&
          h(
            "picture",
            {
              class: pictureClassName,
              style: {
                ...pictureStyle,
                ...absolutePositioning,
                opacity: showImage ? 1 : 0,
                transition:
                  !fadeInDuration || fadeInDuration > 0
                    ? `opacity ${fadeInDuration || 500}ms`
                    : null
              }
            },
            [
              webpSource,
              regularSource,
              data.src &&
                h("img", {
                  style: { maxWidth: "100%" },
                  attrs: {
                    src: data.src,
                    alt: data.alt,
                    title: data.title
                  },
                  on: {
                    load: this.load
                  }
                })
            ]
          ),
        h("noscript", [
          h(
            "picture",
            {
              class: pictureClassName,
              style: { ...pictureStyle, ...absolutePositioning }
            },
            [
              webpSource,
              regularSource,
              data.src &&
                h("img", {
                  attrs: {
                    src: data.src,
                    alt: data.alt,
                    title: data.title
                  }
                })
            ]
          )
        ])
      ]
    );
  },
  mounted() {
    if ("IntersectionObserver" in window) {
      console.log('carico');
      this.observer = new IntersectionObserver(
        entries => {
          console.log('EMITTED!');
          const image = entries[0];
          if (image.isIntersecting) {
            console.log('INTERSE');
            this.inView = true;
            this.observer.disconnect();
          }
        },
        {
          threshold: this.intersectionTreshold,
          rootMargin: this.intersectionMargin
        }
      );
      this.observer.observe(this.$el);
    }
  },
  destroyed() {
    if ("IntersectionObserver" in window && this.observer) {
      this.observer.disconnect();
    }
  }
};

export default Image;

export const DatocmsImagePlugin = {
  install: Vue => {
    Vue.component("DatocmsImage", Image);
  }
};
