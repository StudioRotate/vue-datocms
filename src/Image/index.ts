import {
  defineComponent,
  ref,
  onMounted,
  PropType,
  onBeforeUnmount,
  h,
} from "vue-demi";

export type ResponsiveImageType = {
  /** The aspect ratio (width/height) of the image */
  aspectRatio: number;
  /** A base64-encoded thumbnail to offer during image loading */
  base64?: string;
  /** The height of the image */
  height?: number;
  /** The width of the image */
  width: number;
  /** The HTML5 `sizes` attribute for the image */
  sizes?: string;
  /** The fallback `src` attribute for the image */
  src?: string;
  /** The HTML5 `srcSet` attribute for the image */
  srcSet?: string;
  /** The HTML5 `srcSet` attribute for the image in WebP format, for browsers that support the format */
  webpSrcSet?: string;
  /** The background color for the image placeholder */
  bgColor?: string;
  /** Alternate text (`alt`) for the image */
  alt?: string;
  /** Title attribute (`title`) for the image */
  title?: string;
};

const isSsr = typeof window === "undefined";

const isIntersectionObserverAvailable = isSsr
  ? false
  : !!(window as any).IntersectionObserver;

const universalBtoa = isSsr
  ? (str: string) => Buffer.from(str.toString(), "binary").toString("base64")
  : window.btoa;

const absolutePositioning = {
  position: "absolute",
  left: "0px",
  top: "0px",
  width: "100%",
  height: "100%",
};

const useInView = ({ threshold, rootMargin }: IntersectionObserverInit) => {
  const observer = ref<IntersectionObserver | null>(null);
  const elRef = ref<HTMLElement | null>(null);
  const inView = ref(false);

  onMounted(() => {
    if (isIntersectionObserverAvailable) {
      observer.value = new IntersectionObserver(
        (entries) => {
          const image = entries[0];
          if (image.isIntersecting && observer.value) {
            inView.value = true;
            observer.value.disconnect();
          }
        },
        {
          threshold,
          rootMargin,
        }
      );
      if (elRef.value) {
        observer.value.observe(elRef.value);
      }
    }
  });

  onBeforeUnmount(() => {
    if (isIntersectionObserverAvailable && observer.value) {
      observer.value.disconnect();
    }
  });

  return { inView, elRef };
};

type State = {
  lazyLoad?: boolean;
  inView: boolean;
  loaded: boolean;
};

const imageAddStrategy = ({ lazyLoad, inView, loaded }: State) => {
  if (!lazyLoad) {
    return true;
  }

  if (isSsr) {
    return false;
  }

  if (isIntersectionObserverAvailable) {
    return inView || loaded;
  }

  return true;
};

const imageShowStrategy = ({ lazyLoad, loaded }: State) => {
  if (!lazyLoad) {
    return true;
  }

  if (isSsr) {
    return false;
  }

  if (isIntersectionObserverAvailable) {
    return loaded;
  }

  return true;
};

export const Image = defineComponent({
  name: "DatocmsImage",
  props: {
    /** The actual response you get from a DatoCMS `responsiveImage` GraphQL query */
    data: {
      type: Object as PropType<ResponsiveImageType>,
      required: true,
    },
    /** Additional CSS class for the image inside the `<picture />` tag */
    pictureClass: {
      type: String,
    },
    /** Duration (in ms) of the fade-in transition effect upoad image loading */
    fadeInDuration: {
      type: Number,
    },
    /** @deprecated Use the intersectionThreshold prop */
    intersectionTreshold: {
      type: Number,
      default: 0,
    },
    /** Indicate at what percentage of the placeholder visibility the loading of the image should be triggered. A value of 0 means that as soon as even one pixel is visible, the callback will be run. A value of 1.0 means that the threshold isn't considered passed until every pixel is visible */
    intersectionThreshold: {
      type: Number,
    },
    /** Margin around the placeholder. Can have values similar to the CSS margin property (top, right, bottom, left). The values can be percentages. This set of values serves to grow or shrink each side of the placeholder element's bounding box before computing intersections */
    intersectionMargin: {
      type: String,
      default: "0px 0px 0px 0px",
    },
    /** Wheter enable lazy loading or not */
    lazyLoad: {
      type: Boolean,
      default: true,
    },
    /** Additional CSS rules to add to the image inside the `<picture />` tag */
    pictureStyle: {
      type: Object,
      default: {},
    },
    /** Wheter the image wrapper should explicitely declare the width of the image or keep it fluid */
    explicitWidth: {
      type: Boolean,
    },
  },
  setup(props) {
    const loaded = ref(false);

    function handleLoad() {
      loaded.value = true;
    }

    const { inView, elRef } = useInView({
      threshold: props.intersectionThreshold || props.intersectionTreshold || 0,
      rootMargin: props.intersectionMargin || "0px 0px 0px 0px",
    });

    return () => {
      const addImage = imageAddStrategy({
        lazyLoad: props.lazyLoad,
        inView: inView.value,
        loaded: loaded.value,
      });

      const showImage = imageShowStrategy({
        lazyLoad: props.lazyLoad,
        inView: inView.value,
        loaded: loaded.value,
      });

      const webpSource =
        props.data.webpSrcSet &&
        h("source", {
          srcset: props.data.webpSrcSet,
          sizes: props.data.sizes,
          type: "image/webp",
        });

      const regularSource =
        props.data.srcSet &&
        h("source", {
          srcset: props.data.srcSet,
          sizes: props.data.sizes,
        });

      const transition =
        typeof props.fadeInDuration === "undefined" || props.fadeInDuration > 0
          ? `opacity ${props.fadeInDuration || 500}ms ${
              props.fadeInDuration || 500
            }ms`
          : undefined;

      const placeholder = h("div", {
        style: {
          backgroundImage: props.data.base64
            ? `url(${props.data.base64})`
            : null,
          backgroundColor: props.data.bgColor,
          backgroundSize: "cover",
          opacity: showImage ? 0 : 1,
          transition: transition,
          ...absolutePositioning,
        },
      });

      const { width, aspectRatio } = props.data;

      const height = props.data.height || width / aspectRatio;
      const svg = `<svg xmlns="http://www.w3.org/2000/svg" width="${width}" height="${height}"></svg>`;

      const sizer = h("img", {
        class: props.pictureClass,
        style: {
          display: "block",
          width: props.explicitWidth ? `${width}px` : "100%",
          ...props.pictureStyle,
        },
        src: `data:image/svg+xml;base64,${universalBtoa(svg)}`,
        role: "presentation",
      });

      return h(
        "div",
        {
          style: {
            display: props.explicitWidth ? "inline-block" : "block",
            overflow: "hidden",
            position: "relative",
          },
          ref: elRef,
        },
        [
          sizer,
          placeholder,
          addImage &&
            h("picture", null, [
              webpSource,
              regularSource,
              props.data.src &&
                h("img", {
                  src: props.data.src,
                  alt: props.data.alt,
                  title: props.data.title,
                  onLoad: handleLoad,
                  class: props.pictureClass,
                  style: {
                    ...absolutePositioning,
                    ...props.pictureStyle,
                    opacity: showImage ? 1 : 0,
                    transition,
                  },
                }),
            ]),
          h("noscript", null, [
            h("picture", null, [
              webpSource,
              regularSource,
              props.data.src &&
                h("img", {
                  src: props.data.src,
                  alt: props.data.alt,
                  title: props.data.title,
                  class: props.pictureClass,
                  style: {
                    ...props.pictureStyle,
                    ...absolutePositioning,
                  },
                  loading: "lazy",
                }),
            ]),
          ]),
        ]
      );
    };
  },
});

export const DatocmsImagePlugin = {
  install: (Vue: any) => {
    Vue.component("DatocmsImage", Image);
  },
};
