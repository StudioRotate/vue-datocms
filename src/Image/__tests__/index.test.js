import { mount } from "@vue/test-utils";
import Image from "../";

const data = {
  alt: "DatoCMS swag",
  aspectRatio: 1.7777777777777777,
  base64:
    "data:image/jpeg;base64,/9j/4AAQSkZJRgABAQAAAQABAAD/2wCEAAoHBwgHBgoICAgLFQoLDhgQDg0NDh0eHREYIx8lJCIrHB0dLSs7GikyKSEuKjUlKDk1MjIyHyo4PTc+PDcxPjUBCgsLDg0OHBAQHDsoIig7Ozs7Ozs7OzsvOzs7Ozs7Ozs7Lzs7Ozs7Ozs7OzsvOzs7NTsvLy87NTU1Ly8vLzsvL//AABEIAA0AGAMBIgACEQEDEQH/xAAYAAACAwAAAAAAAAAAAAAAAAAGBwABBP/EACEQAAEEAAYDAAAAAAAAAAAAAAEAAgMEBQYHESEiFWFx/8QAFQEBAQAAAAAAAAAAAAAAAAAAAwL/xAAZEQADAAMAAAAAAAAAAAAAAAAAAQIRITH/2gAMAwEAAhEDEQA/AFxLgDWTsAd1J5TGy7hEYqNAaNgECX7sjLMQAHJTEy1Zcarfia4lJMauAxqBhLY6ZlaOzDurWvUOd3jZPfCiEh4xs//Z",
  height: 421,
  sizes: "(max-width: 750px) 100vw, 750px",
  src: "https://www.datocms-assets.com/205/image.png?ar=16%3A9&fit=crop&w=750",
  srcSet:
    "https://www.datocms-assets.com/205/image.png?ar=16%3A9&dpr=0.25&fit=crop&w=750 187w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&dpr=0.5&fit=crop&w=750 375w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&dpr=0.75&fit=crop&w=750 562w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&dpr=1&fit=crop&w=750 750w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&dpr=1.5&fit=crop&w=750 1125w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&dpr=2&fit=crop&w=750 1500w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&dpr=3&fit=crop&w=750 2250w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&dpr=4&fit=crop&w=750 3000w",
  title: "These are awesome, we know that.",
  width: 750
};

const observerMap = new Map();
const instanceMap = new Map();

describe("Image", () => {
  beforeAll(() => {
    global.IntersectionObserver = jest.fn((cb, options) => {
      const instance = {
        thresholds: Array.isArray(options.threshold)
          ? options.threshold
          : [options.threshold],
        root: options.root,
        rootMargin: options.rootMargin,
        time: Date.now(),
        observe: jest.fn(element => {
          instanceMap.set(element, instance);
          observerMap.set(element, cb);
        }),
        unobserve: jest.fn(element => {
          instanceMap.delete(element);
          observerMap.delete(element);
        }),
        disconnect: jest.fn()
      };
      return instance;
    });
  });

  afterEach(() => {
    global.IntersectionObserver.mockClear();
    instanceMap.clear();
    observerMap.clear();
  });

  function mockAllIsIntersecting(isIntersecting) {
    observerMap.forEach((onChange, element) => {
      mockIsIntersecting(element, isIntersecting)
    })
  }

  function mockIsIntersecting(element, isIntersecting) {
    const cb = observerMap.get(element)
    const instance = instanceMap.get(element)
    if (cb && instance) {
      const entry = [
        {
          boundingClientRect: element.getBoundingClientRect(),
          intersectionRatio: isIntersecting ? 1 : 0,
          intersectionRect: isIntersecting ? element.getBoundingClientRect() : {},
          isIntersecting,
          rootBounds: instance.root ? instance.root.getBoundingClientRect() : {},
          target: element,
          time: Date.now() - instance.time,
        },
      ]
      cb(entry, instance)
    } else {
      throw new Error(
        'No IntersectionObserver instance found for element. Is it still mounted in the DOM?',
      )
    }
  }

  describe("not visible", () => {
    it("renders the blur-up thumb", async () => {
      const wrapper = mount(Image, {
        propsData: {
          data
        }
      });
      await wrapper.vm.$nextTick();
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe("visible", () => {
    it("renders the image", async () => {
      const wrapper = mount(Image, {
        propsData: {
          data
        }
      });
      mockAllIsIntersecting(true);
      await wrapper.vm.$nextTick();
      expect(wrapper).toMatchSnapshot();
    });

    describe("image loaded", () => {
      it("shows the image", async () => {
        const wrapper = mount(Image, {
          propsData: {
            data
          }
        });
        mockAllIsIntersecting(true);
        await wrapper.vm.$nextTick();
        wrapper.find("img").trigger("load");
        await wrapper.vm.$nextTick();
        expect(wrapper).toMatchSnapshot();
      });
    });
  });
});
