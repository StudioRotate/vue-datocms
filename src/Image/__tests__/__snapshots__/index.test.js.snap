// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image not visible renders the blur-up thumb 1`] = `
<div style="display: inline-block; overflow: hidden; position: relative;"><svg height="421" width="750" style="width: 100%; height: auto; display: block;"></svg>
  <div style="background-image: url(data:image/jpeg;base64,/9j/4AAQSkZJRgABAQAAAQABAAD/2wCEAAoHBwgHBgoICAgLFQoLDhgQDg0NDh0eHREYIx8lJCIrHB0dLSs7GikyKSEuKjUlKDk1MjIyHyo4PTc+PDcxPjUBCgsLDg0OHBAQHDsoIig7Ozs7Ozs7OzsvOzs7Ozs7Ozs7Lzs7Ozs7Ozs7OzsvOzs7NTsvLy87NTU1Ly8vLzsvL//AABEIAA0AGAMBIgACEQEDEQH/xAAYAAACAwAAAAAAAAAAAAAAAAAGBwABBP/EACEQAAEEAAYDAAAAAAAAAAAAAAEAAgMEBQYHESEiFWFx/8QAFQEBAQAAAAAAAAAAAAAAAAAAAwL/xAAZEQADAAMAAAAAAAAAAAAAAAAAAQIRITH/2gAMAwEAAhEDEQA/AFxLgDWTsAd1J5TGy7hEYqNAaNgECX7sjLMQAHJTEy1Zcarfia4lJMauAxqBhLY6ZlaOzDurWvUOd3jZPfCiEh4xs//Z); background-size: cover; opacity: 1; transition: opacity 500ms 500ms; position: absolute; left: 0px; top: 0px; right: 0px; bottom: 0px;"></div><noscript>
    <picture style="position: absolute; left: 0px; top: 0px; right: 0px; bottom: 0px; ">
      <source srcset="https://www.datocms-assets.com/205/image.png?ar=16%3A9&amp;dpr=0.25&amp;fit=crop&amp;w=750 187w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&amp;dpr=0.5&amp;fit=crop&amp;w=750 375w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&amp;dpr=0.75&amp;fit=crop&amp;w=750 562w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&amp;dpr=1&amp;fit=crop&amp;w=750 750w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&amp;dpr=1.5&amp;fit=crop&amp;w=750 1125w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&amp;dpr=2&amp;fit=crop&amp;w=750 1500w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&amp;dpr=3&amp;fit=crop&amp;w=750 2250w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&amp;dpr=4&amp;fit=crop&amp;w=750 3000w" sizes="(max-width: 750px) 100vw, 750px" /><img loading="lazy" src="https://www.datocms-assets.com/205/image.png?ar=16%3A9&amp;fit=crop&amp;w=750" alt="DatoCMS swag" title="These are awesome, we know that." /></picture>
  </noscript>
</div>
`;

exports[`Image visible image loaded shows the image 1`] = `
<div style="display: inline-block; overflow: hidden; position: relative;"><svg height="421" width="750" style="width: 100%; height: auto; display: block;"></svg>
  <div style="background-image: url(data:image/jpeg;base64,/9j/4AAQSkZJRgABAQAAAQABAAD/2wCEAAoHBwgHBgoICAgLFQoLDhgQDg0NDh0eHREYIx8lJCIrHB0dLSs7GikyKSEuKjUlKDk1MjIyHyo4PTc+PDcxPjUBCgsLDg0OHBAQHDsoIig7Ozs7Ozs7OzsvOzs7Ozs7Ozs7Lzs7Ozs7Ozs7OzsvOzs7NTsvLy87NTU1Ly8vLzsvL//AABEIAA0AGAMBIgACEQEDEQH/xAAYAAACAwAAAAAAAAAAAAAAAAAGBwABBP/EACEQAAEEAAYDAAAAAAAAAAAAAAEAAgMEBQYHESEiFWFx/8QAFQEBAQAAAAAAAAAAAAAAAAAAAwL/xAAZEQADAAMAAAAAAAAAAAAAAAAAAQIRITH/2gAMAwEAAhEDEQA/AFxLgDWTsAd1J5TGy7hEYqNAaNgECX7sjLMQAHJTEy1Zcarfia4lJMauAxqBhLY6ZlaOzDurWvUOd3jZPfCiEh4xs//Z); background-size: cover; opacity: 0; transition: opacity 500ms 500ms; position: absolute; left: 0px; top: 0px; right: 0px; bottom: 0px;"></div>
  <picture style="position: absolute; left: 0px; top: 0px; right: 0px; bottom: 0px; opacity: 1; transition: opacity 500ms;">
    <source srcset="https://www.datocms-assets.com/205/image.png?ar=16%3A9&amp;dpr=0.25&amp;fit=crop&amp;w=750 187w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&amp;dpr=0.5&amp;fit=crop&amp;w=750 375w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&amp;dpr=0.75&amp;fit=crop&amp;w=750 562w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&amp;dpr=1&amp;fit=crop&amp;w=750 750w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&amp;dpr=1.5&amp;fit=crop&amp;w=750 1125w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&amp;dpr=2&amp;fit=crop&amp;w=750 1500w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&amp;dpr=3&amp;fit=crop&amp;w=750 2250w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&amp;dpr=4&amp;fit=crop&amp;w=750 3000w" sizes="(max-width: 750px) 100vw, 750px"><img src="https://www.datocms-assets.com/205/image.png?ar=16%3A9&amp;fit=crop&amp;w=750" alt="DatoCMS swag" title="These are awesome, we know that." style="width: 100%;"></picture><noscript>
    <picture style="position: absolute; left: 0px; top: 0px; right: 0px; bottom: 0px; ">
      <source srcset="https://www.datocms-assets.com/205/image.png?ar=16%3A9&amp;dpr=0.25&amp;fit=crop&amp;w=750 187w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&amp;dpr=0.5&amp;fit=crop&amp;w=750 375w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&amp;dpr=0.75&amp;fit=crop&amp;w=750 562w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&amp;dpr=1&amp;fit=crop&amp;w=750 750w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&amp;dpr=1.5&amp;fit=crop&amp;w=750 1125w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&amp;dpr=2&amp;fit=crop&amp;w=750 1500w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&amp;dpr=3&amp;fit=crop&amp;w=750 2250w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&amp;dpr=4&amp;fit=crop&amp;w=750 3000w" sizes="(max-width: 750px) 100vw, 750px" /><img loading="lazy" src="https://www.datocms-assets.com/205/image.png?ar=16%3A9&amp;fit=crop&amp;w=750" alt="DatoCMS swag" title="These are awesome, we know that." /></picture>
  </noscript>
</div>
`;

exports[`Image visible renders the image 1`] = `
<div style="display: inline-block; overflow: hidden; position: relative;"><svg height="421" width="750" style="width: 100%; height: auto; display: block;"></svg>
  <div style="background-image: url(data:image/jpeg;base64,/9j/4AAQSkZJRgABAQAAAQABAAD/2wCEAAoHBwgHBgoICAgLFQoLDhgQDg0NDh0eHREYIx8lJCIrHB0dLSs7GikyKSEuKjUlKDk1MjIyHyo4PTc+PDcxPjUBCgsLDg0OHBAQHDsoIig7Ozs7Ozs7OzsvOzs7Ozs7Ozs7Lzs7Ozs7Ozs7OzsvOzs7NTsvLy87NTU1Ly8vLzsvL//AABEIAA0AGAMBIgACEQEDEQH/xAAYAAACAwAAAAAAAAAAAAAAAAAGBwABBP/EACEQAAEEAAYDAAAAAAAAAAAAAAEAAgMEBQYHESEiFWFx/8QAFQEBAQAAAAAAAAAAAAAAAAAAAwL/xAAZEQADAAMAAAAAAAAAAAAAAAAAAQIRITH/2gAMAwEAAhEDEQA/AFxLgDWTsAd1J5TGy7hEYqNAaNgECX7sjLMQAHJTEy1Zcarfia4lJMauAxqBhLY6ZlaOzDurWvUOd3jZPfCiEh4xs//Z); background-size: cover; opacity: 1; transition: opacity 500ms 500ms; position: absolute; left: 0px; top: 0px; right: 0px; bottom: 0px;"></div>
  <picture style="position: absolute; left: 0px; top: 0px; right: 0px; bottom: 0px; opacity: 0; transition: opacity 500ms;">
    <source srcset="https://www.datocms-assets.com/205/image.png?ar=16%3A9&amp;dpr=0.25&amp;fit=crop&amp;w=750 187w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&amp;dpr=0.5&amp;fit=crop&amp;w=750 375w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&amp;dpr=0.75&amp;fit=crop&amp;w=750 562w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&amp;dpr=1&amp;fit=crop&amp;w=750 750w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&amp;dpr=1.5&amp;fit=crop&amp;w=750 1125w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&amp;dpr=2&amp;fit=crop&amp;w=750 1500w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&amp;dpr=3&amp;fit=crop&amp;w=750 2250w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&amp;dpr=4&amp;fit=crop&amp;w=750 3000w" sizes="(max-width: 750px) 100vw, 750px"><img src="https://www.datocms-assets.com/205/image.png?ar=16%3A9&amp;fit=crop&amp;w=750" alt="DatoCMS swag" title="These are awesome, we know that." style="width: 100%;"></picture><noscript>
    <picture style="position: absolute; left: 0px; top: 0px; right: 0px; bottom: 0px; ">
      <source srcset="https://www.datocms-assets.com/205/image.png?ar=16%3A9&amp;dpr=0.25&amp;fit=crop&amp;w=750 187w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&amp;dpr=0.5&amp;fit=crop&amp;w=750 375w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&amp;dpr=0.75&amp;fit=crop&amp;w=750 562w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&amp;dpr=1&amp;fit=crop&amp;w=750 750w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&amp;dpr=1.5&amp;fit=crop&amp;w=750 1125w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&amp;dpr=2&amp;fit=crop&amp;w=750 1500w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&amp;dpr=3&amp;fit=crop&amp;w=750 2250w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&amp;dpr=4&amp;fit=crop&amp;w=750 3000w" sizes="(max-width: 750px) 100vw, 750px" /><img loading="lazy" src="https://www.datocms-assets.com/205/image.png?ar=16%3A9&amp;fit=crop&amp;w=750" alt="DatoCMS swag" title="These are awesome, we know that." /></picture>
  </noscript>
</div>
`;
